windows_exporter integration: Panic from MSSQL collector

After Updating to the later versions of the Agent. v0.20+ the configuration file that i was using no longer works and is causing the agent to not run.

The relevant section of interest from the config file is as follows:

```yaml
integrations:
  windows_exporter:
    enabled: true 
    enabled_collectors: cpu,cs,logical_disk,net,os,service,system,textfile,mssql
```

This works correctly for v0.19 but any later version will panic with the following:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x3406dad]

goroutine 371 [running]:
github.com/prometheus-community/windows_exporter/collector.(*MSSQLCollector).Col
lect(0xc000bcd800, 0xc0005a71c0, 0xc000d45ce0, 0x0, 0x0)
/go/pkg/mod/github.com/grafana/windows_exporter@v0.15.1-0.20211019183116
-592dfa92f9fd/collector/mssql.go:1861 +0x6d
github.com/prometheus-community/windows_exporter/collector.execute(0xc000840a33,
0x5, 0x49ca018, 0xc000bcd800, 0xc0005a71c0, 0xc000d45ce0, 0x0)
/go/pkg/mod/github.com/grafana/windows_exporter@v0.15.1-0.20211019183116
-592dfa92f9fd/collector/prometheus.go:173 +0x8f
github.com/prometheus-community/windows_exporter/collector.(*Prometheus).Collect
.func2(0xc00111d3f0, 0xc0005a71c0, 0xc000d45ce0, 0xc00111d400, 0xc00111d408, 0xc
000ec57a0, 0xc000840a33, 0x5, 0x49ca018, 0xc000bcd800)
/go/pkg/mod/github.com/grafana/windows_exporter@v0.15.1-0.20211019183116
-592dfa92f9fd/collector/prometheus.go:114 +0x9c
created by github.com/prometheus-community/windows_exporter/collector.(*Promethe
us).Collect
/go/pkg/mod/github.com/grafana/windows_exporter@v0.15.1-0.20211019183116
-592dfa92f9fd/collector/prometheus.go:112 +0x64a
```

This is from a fresh install of the 21.2 Agent with Windows Exporter Enabled during the install

This appears to be the output it gives

c:\Program Files\Grafana Agent v0.21.2>agent-windows-amd64.exe -config.file "C:\
Program Files\Grafana Agent v0.21.2\agent-config.yaml"
ts=2021-12-15T23:19:08.8160029Z caller=node.go:85 level=info agent=prometheus co
mponent=cluster msg="applying config"
ts=2021-12-15T23:19:08.8169778Z caller=remote.go:180 level=info agent=prometheus
 component=cluster msg="not watching the KV, none set"
ts=2021-12-15T23:19:08Z level=info caller=traces/traces.go:123 msg="Traces Logge
r Initialized" component=traces
time="2021-12-15T23:19:08Z" level=warning msg="No where-clause specified for ser
vice collector. This will generate a very large number of metrics!" source="serv
ice.go:49"
ts=2021-12-15T23:19:08.8325762Z caller=windows_exporter_windows.go:32 level=info
 integration=windows_exporter msg="enabled windows_exporter collectors" collecto
rs="unsupported value type"
ts=2021-12-15T23:19:08.8335511Z caller=wal.go:182 level=info agent=prometheus in
stance=13e6fc0b99e0f744ea97242498af5c1a msg="replaying WAL, this may take a whil
e" dir=C:\ProgramData\grafana-agent-wal\13e6fc0b99e0f744ea97242498af5c1a\wal
ts=2021-12-15T23:19:08.834526Z caller=wal.go:229 level=info agent=prometheus ins
tance=13e6fc0b99e0f744ea97242498af5c1a msg="WAL segment loaded" segment=0 maxSeg
ment=2
ts=2021-12-15T23:19:08.8374507Z caller=wal.go:229 level=info agent=prometheus in
stance=13e6fc0b99e0f744ea97242498af5c1a msg="WAL segment loaded" segment=1 maxSe
gment=2
ts=2021-12-15T23:19:08.8433001Z caller=wal.go:229 level=info agent=prometheus in
stance=13e6fc0b99e0f744ea97242498af5c1a msg="WAL segment loaded" segment=2 maxSe
gment=2
ts=2021-12-15T23:19:08.8471997Z caller=dedupe.go:112 agent=prometheus instance=1
3e6fc0b99e0f744ea97242498af5c1a component=remote level=info remote_name=13e6fc-3
20289 url=https://prometheus-us-central1.grafana.net/api/prom/push msg="Starting
 WAL watcher" queue=13e6fc-320289
ts=2021-12-15T23:19:08.8471997Z caller=dedupe.go:112 agent=prometheus instance=1
3e6fc0b99e0f744ea97242498af5c1a component=remote level=info remote_name=13e6fc-3
20289 url=https://prometheus-us-central1.grafana.net/api/prom/push msg="Starting
 scraped metadata watcher"
ts=2021-12-15T23:19:08.8471997Z caller=dedupe.go:112 agent=prometheus instance=1
3e6fc0b99e0f744ea97242498af5c1a component=remote level=info remote_name=13e6fc-3
20289 url=https://prometheus-us-central1.grafana.net/api/prom/push msg="Replayin
g WAL" queue=13e6fc-320289
ts=2021-12-15T23:19:08.8491495Z caller=dedupe.go:112 agent=prometheus instance=1
3e6fc0b99e0f744ea97242498af5c1a component=remote level=info remote_name=13e6fc-5
a3fdf url=https://auxprom01.ukwest.cloudapp.azure.com:5721/api/v1/write msg="Sta
rting WAL watcher" queue=13e6fc-5a3fdf
ts=2021-12-15T23:19:08.8491495Z caller=dedupe.go:112 agent=prometheus instance=1
3e6fc0b99e0f744ea97242498af5c1a component=remote level=info remote_name=13e6fc-5
a3fdf url=https://auxprom01.ukwest.cloudapp.azure.com:5721/api/v1/write msg="Sta
rting scraped metadata watcher"
ts=2021-12-15T23:19:08.8501244Z caller=dedupe.go:112 agent=prometheus instance=1
3e6fc0b99e0f744ea97242498af5c1a component=remote level=info remote_name=13e6fc-5
a3fdf url=https://auxprom01.ukwest.cloudapp.azure.com:5721/api/v1/write msg="Rep
laying WAL" queue=13e6fc-5a3fdf
ts=2021-12-15T23:19:08.8657228Z caller=manager.go:194 level=info agent=prometheu
s msg="dynamically updated instance" instance=13e6fc0b99e0f744ea97242498af5c1a
ts=2021-12-15T23:19:08.8666977Z caller=server.go:77 level=info msg="server confi
guration changed, restarting server"
ts=2021-12-15T23:19:08.8686475Z caller=gokit.go:72 level=info http=[::]:12345 gr
pc=[::]:9095 msg="server listening on addresses"
time="2021-12-15T23:19:31Z" level=error msg="Error reading textfile collector di
rectory \"C:\\\\Program Files\\\\windows_exporter\\\\textfile_inputs\": open C:\
\Program Files\\windows_exporter\\textfile_inputs: The system cannot find the pa
th specified." source="textfile.go:235"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x3406dad]

goroutine 371 [running]:
github.com/prometheus-community/windows_exporter/collector.(*MSSQLCollector).Col
lect(0xc000bcd800, 0xc0005a71c0, 0xc000d45ce0, 0x0, 0x0)
        /go/pkg/mod/github.com/grafana/windows_exporter@v0.15.1-0.20211019183116
-592dfa92f9fd/collector/mssql.go:1861 +0x6d
github.com/prometheus-community/windows_exporter/collector.execute(0xc000840a33,
 0x5, 0x49ca018, 0xc000bcd800, 0xc0005a71c0, 0xc000d45ce0, 0x0)
        /go/pkg/mod/github.com/grafana/windows_exporter@v0.15.1-0.20211019183116
-592dfa92f9fd/collector/prometheus.go:173 +0x8f
github.com/prometheus-community/windows_exporter/collector.(*Prometheus).Collect
.func2(0xc00111d3f0, 0xc0005a71c0, 0xc000d45ce0, 0xc00111d400, 0xc00111d408, 0xc
000ec57a0, 0xc000840a33, 0x5, 0x49ca018, 0xc000bcd800)
        /go/pkg/mod/github.com/grafana/windows_exporter@v0.15.1-0.20211019183116
-592dfa92f9fd/collector/prometheus.go:114 +0x9c
created by github.com/prometheus-community/windows_exporter/collector.(*Promethe
us).Collect
        /go/pkg/mod/github.com/grafana/windows_exporter@v0.15.1-0.20211019183116
-592dfa92f9fd/collector/prometheus.go:112 +0x64a

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -25,6 +25,8 @@
 - [BUGFIX] integrations-next: Wait for integrations to stop when starting new
   instances or shutting down (@rfratto).
 
+- [BUGFIX] Fix issue with windows_exporter mssql collector crashing the agent. (@mattdurham)
+
 # v0.22.0 (2022-01-13)
 
 This release has deprecations. Please read [DEPRECATION] entries and consult

--- a/go.mod
+++ b/go.mod
@@ -395,7 +395,7 @@ replace (
 	github.com/percona/mongodb_exporter => github.com/grafana/mongodb_exporter v0.20.8-0.20211006135645-bef0f0239601
 	github.com/percona/percona-toolkit => github.com/rlankfo/percona-toolkit v0.0.0-20211028191359-7aada1bf148f
 	github.com/prometheus-community/postgres_exporter => github.com/grafana/postgres_exporter v0.8.1-0.20210722175051-db35d7c2f520
-	github.com/prometheus-community/windows_exporter => github.com/grafana/windows_exporter v0.15.1-0.20211019183116-592dfa92f9fd
+	github.com/prometheus-community/windows_exporter => github.com/grafana/windows_exporter v0.15.1-0.20220202211901-871715ba0b43
 	github.com/prometheus/mysqld_exporter => github.com/grafana/mysqld_exporter v0.12.2-0.20201015182516-5ac885b2d38a
 	github.com/prometheus/statsd_exporter => github.com/grafana/statsd_exporter v0.18.1-0.20211118164740-8e806158da0b
 )
@@ -421,3 +421,5 @@ replace github.com/jaegertracing/jaeger => github.com/jaegertracing/jaeger v1.27
 
 // Replacement necessary for windows_exporter so that we can use gokit logging and not the old prometheus logging
 replace github.com/leoluk/perflib_exporter v0.1.0 => github.com/grafana/perflib_exporter v0.1.1-0.20211013152516-e37e14fb8b0a
+
+

--- a/pkg/integrations/windows_exporter/windows_exporter_windows.go
+++ b/pkg/integrations/windows_exporter/windows_exporter_windows.go
@@ -29,7 +29,7 @@ func New(log log.Logger, c *Config) (integrations.Integration, error) {
 		collectorNames = append(collectorNames, key)
 	}
 	sort.Strings(collectorNames)
-	level.Info(log).Log("msg", "enabled windows_exporter collectors", "collectors", collectorNames)
+	level.Info(log).Log("msg", "enabled windows_exporter collectors", "collectors", strings.Join(collectorNames, ","))
 
 	return integrations.NewCollectorIntegration(c.Name(), integrations.WithCollectors(
 		// Hard-coded 4m timeout to represent the time a series goes stale.

